File: mc/rmcp/src/main/java/calebxzhou/rdi/mc/common2/mcp/RMcpRecipeData.java
package calebxzhou.rdi.mc.common2.mcp;

import java.util.List;
import java.util.Map;

public record RMcpRecipeData(
        String id,
        String source,
        String type,
        String category,
        String title,
        String runtimeClass,
        List<IngredientSlot> inputs,
        List<IngredientSlot> outputs,
        List<IngredientSlot> catalysts,
        List<IngredientSlot> renderOnly,
        Map<String, Object> extra
) {
    public record IngredientSlot(
            String role,
            List<Item> items,
            List<Fluid> fluids,
            List<ItemTag> tags
    ) {
        public IngredientSlot(String role, List<Item> items, List<Fluid> fluids) {
            this(role, items, fluids, List.of());
        }
    }

    public record ItemTag(
            String id,
            int count,
            int candidateCount,
            List<Item> examples,
            String source
    ) {
    }

    public record Item(
            String id,
            String langKey,
            int count,
            String snbt
    ) {
    }

    public record Fluid(
            String id,
            String name,
            long amount
    ) {
    }
}

File: mc/rmcp/common/src/main/kotlin/calebxzhou/rdi/mc/common2/mcp/RMcpRecipeData.kt
package calebxzhou.rdi.mc.common2.mcp

@JvmRecord
data class RMcpRecipeData(
    val id: String,
    val source: String,
    val type: String,
    val category: String,
    val title: String,
    val runtimeClass: String,
    val inputs: List<IngredientSlot>,
    val outputs: List<IngredientSlot>,
    val catalysts: List<IngredientSlot>,
    val renderOnly: List<IngredientSlot>,
    val extra: Map<String, Any?>?
) {
    @JvmRecord
    data class IngredientSlot(
        val role: String,
        val items: List<Item>,
        val fluids: List<Fluid>,
        val tags: List<ItemTag>
    ) {
        constructor(role: String, items: List<Item>, fluids: List<Fluid>) : this(
            role,
            items,
            fluids,
            mutableListOf<ItemTag>()
        )
    }

    @JvmRecord
    data class ItemTag(
        val id: String,
        val count: Int,
        val candidateCount: Int,
        val examples: List<Item>,
        val source: String?
    )

    @JvmRecord
    data class Item(
        val id: String,
        val langKey: String,
        val count: Int,
        val snbt: String?
    )

    @JvmRecord
    data class Fluid(
        val id: String,
        val name: String,
        val amount: Long
    )
}
